### PR TITLE
Bazel CI

### LINKED ISSUE #264: Bazel + CI
We should run our Bazel build/test/coverage in CI.

There's no way to know if a PR breaks the Bazel support other than manually running it.

### LINKED ISSUE #265: Migrate Docker Nix to Bazel
Drop Nix and use Bazel to build our Docker images.

We lack internal Nix expertise to continue using it.

References
- https://github.com/bazelbuild/rules_docker
- https://github.com/bazelbuild/rules_docker#rust_image

### LINKED ISSUE #294: Migrate BATs tests to Bazel
See https://github.com/filmil/bazel-bats

It takes care of installing bats and the test extensions (assert, support), as well as running the tests.

### PR DESCRIPTION
**Pull request by fmorency**
- CI uses Bazel for lint/build/test/coverage and more.
- Use BuildBuddy Bazel remote cache capabilities
- Use Bazel binaries in Bats tests
- Nuke Nix; build Docker images using Bazel
- Refactor resiliency tests to use Bazel Docker images
- Use `xlarge` Linux Docker executor in CI for everything except Docker-related steps
- Run e2e Bats tests via Bazel
- Run resiliency Bats tests via Bazel
- Enable token test that was not supported by Nix
- Remove staging file duplication
- Make BATs file output TAP-compliant
- bats-assert and bats-support are now handled by Bazel

Fixes #294
Fixes #264
Fixes #265 

TODO: 
- Build Docker images on `macos` (building works, I need a system to test it on!)
- Proper Bazel transitions
- Optimizations
- DRY (macros)

### CODE REVIEW

**Comment by fmorency** [comment]
@hansl I'm not 100% finished and need to merge main into this branch, but I think you can start reviewing

File: BUILD.bazel [hunk 1]
```diff
@@ -1,6 +1,174 @@
...
+        "//docker:openssl_image.tar",
+        "//docker:tendermint_image.tar",
+        "//src/many-abci:many-abci-image.tar",
+        "//src/many-ledger:many-ledger-image.tar",
+    ],
+    tags = ["manual"],
+)
+
+run_make(
+    name = "start-ledger-cluster-dettached",
```
Comment (hansl) [inline BUILD.bazel:92]: typo: `detached` here and everywhere.

File: BUILD.bazel [hunk 2]
```diff
@@ -1,6 +1,174 @@
...
+])
 
 # Required for Vergen
 exports_files([".git"])
+
+## PACKAGE CONFIGURATION ##
+# TODO: Get the version from the Cargo manifest
+VERSION = "0.1.0"
+
+# Exposes the value of the compilation mode to the package naming.
```
Comment (hansl) [inline BUILD.bazel:18]: Can you just add as a comment that this is used below for the `package_file_name` in the `pkg_tar` rule?

File: WORKSPACE.bazel [hunk 3]
```diff
@@ -74,3 +81,160 @@ crates_repository(
...
+    # https://github.com/bazelbuild/rules_docker/issues/1902
+    go_repository_default_config = "@//:WORKSPACE.bazel",
+)
+
+load(
+    "@io_bazel_rules_docker//container:container.bzl",
+    "container_pull",
+)
+
+container_pull(
```
Comment (hansl) [inline WORKSPACE.bazel:170]: Should we use SHA for container_pull to make sure the tag doesn't move?

File: docker/.gitignore [hunk 4]
```diff
@@ -1,2 +1 @@
-*/genfiles/
-e2e/docker-compose.json
\ No newline at end of file
+genfiles-*/
```
Comment (hansl) [inline docker/.gitignore:1]:
```suggestion
genfiles-*/

```

File: docker/Makefile [hunk 5]
```diff
@@ -0,0 +1,36 @@
+SHELL = bash
```
Comment (hansl) [inline docker/Makefile:1]: Maybe as a follow up PR (and issue), can we get rid of Make and use bazel docker rules to start/stop nodes?

File: third_party/cargo/patches/BUILD [hunk 6]
```diff
@@ -0,0 +1,5 @@
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+package(default_visibility = ["//visibility:public"])
```
Comment (hansl) [inline third_party/cargo/patches/BUILD:5]:
```suggestion
package(default_visibility = ["//visibility:public"])

```

File: tests/test_helper/many.bash [hunk 7]
```diff
@@ -55,9 +55,24 @@ function wait_for_block() {
...
+        done >/dev/null
+    done
+}
+
+function wait_for_server() {
+    sleep 30 # Required because of https://github.com/liftedinit/many-rs/issues/307
+    timeout 60s bash -c probe_server "$@"
+}
+
+export -f probe_server
```
Comment (hansl) [inline tests/test_helper/many.bash:78]:
```suggestion
export -f probe_server

```

File: third_party/cargo/patches/BUILD [hunk 8]
```diff
@@ -0,0 +1,5 @@
+# Copyright lowRISC contributors.
```
Comment (hansl) [inline third_party/cargo/patches/BUILD:1]: ?

**Review by hansl — CHANGES_REQUESTED** [review]

File: WORKSPACE.bazel (hunk 3 above)
Comment (fmorency) [reply WORKSPACE.bazel:170]: I'll fix it for Linux and add multi-args capabilities in a future PR.

File: docker/Makefile (hunk 5 above)
Comment (fmorency) [reply docker/Makefile:1]: https://github.com/liftedinit/many-rs/issues/312

File: .circleci/config.yml [hunk 9]
```diff
@@ -341,12 +252,49 @@ commands:
...
+                  # GCC 12 is needed to build `librocksdb-sys` with the `-Wno-error=coverage-invalid-line-number` CXXFLAGS (Bazel)
+                  sudo DEBIAN_FRONTEND=noninteractive apt -y install gcc-12 g++-12
+                  sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-11 11
+                  sudo update-alternatives --install /usr/bin/g++ g++ /usr/bin/g++-11 11
+                  sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-12 12
+                  sudo update-alternatives --install /usr/bin/g++ g++ /usr/bin/g++-12 12
+            - run:
+                name: installing bazelisk
+                command: |
+                  wget https://github.com/bazelbuild/bazelisk/releases/download/v1.15.0/bazelisk-linux-amd64
```
[the remark this answers is not in the document]
Comment (fmorency) [reply .circleci/config.yml]: SHAs are currently unavailable from the Bazelisk GitHub release page. I asked on their Slack.

File: .circleci/config.yml [hunk 10]
```diff
@@ -341,12 +252,49 @@ commands:
...
+                  chmod +x bazelisk-linux-amd64 
+                  sudo mv bazelisk-linux-amd64 /usr/local/bin/bazel
+      - when:
+          condition:
+            equal: [ "macos", << parameters.os >> ]
+          steps:
+            - run:
+                name: installing bazelisk
+                command: |
+                  curl -L -O https://github.com/bazelbuild/bazelisk/releases/download/v1.16.0/bazelisk-darwin-amd64
```
[the remark this answers is not in the document]
Comment (fmorency) [reply .circleci/config.yml:280]: It takes forever to run `brew` on the CI (I stopped it after 2mins) vs fetching the binary (2s).

Let's stick to the current method.

**Review by hansl — APPROVED** [review]

**Pull request merged**